queue/priority_queue.go

```diff
@@ -0,0 +1,65 @@
+package queue
+
+import (
+	"container/heap"
+)
+
+type PriorityQueueItem interface {
+	Less(other PriorityQueueItem) bool
+}
+
+type priorityQueueRepr []PriorityQueueItem
+
+// Len returns the length of the ExpiryQueue.
+func (eq priorityQueueRepr) Len() int { return len(eq) }
+
+// Less is used to order ExpiryQueueItem's by their expiry such that items with
+// the oldest expiry are less than items with newer expiries.
+func (eq priorityQueueRepr) Less(i, j int) bool {
+	return eq[i].Less(eq[j])
+}
+
+// Swap swaps two items in the ExpiryQueue. Swap is used by heap.Interface.
+func (eq priorityQueueRepr) Swap(i, j int) {
+	eq[i], eq[j] = eq[j], eq[i]
+}
+
+// Push adds a new item the the ExpiryQueue.
+func (eq *priorityQueueRepr) Push(x interface{}) {
+	item := x.(PriorityQueueItem)
+	*eq = append(*eq, item)
+}
+
+// Pop removes the top item from the ExpiryQueue. The top item is always the
+// one that holds the next value to expire.
+func (eq *priorityQueueRepr) Pop() interface{} {
+	old := *eq
+	n := len(old)
+	item := old[n-1]
```

Comment:
The final element should be set to nil in order to free it for collection.

Reply:
Done.